### PR TITLE
[clang] Generate nuw GEPs for struct member accesses

### PR DESCRIPTION
**Pull request by hazzlim**
Generate nuw GEPs for struct member accesses, as these are always constant positive offsets derived from the index of the struct member.

Regression tests are updated using update_cc_test_checks.py where possible, and by find + replace where not.

### CODE REVIEW

**Comment by hazzlim** [comment]
This seemed like the right place to make the change - please let me know if not!
 
I've not restricted this to a subset of languages because it seems fairly unambiguous that nuw holds for these inbounds struct member accesses, but it could be limited to e.g. C/C++ if that is more appropriate.

**Review by nikic — COMMENTED** [review]
This looks fine to me. Worth noting though that these nuw flags could also be inferred (inbounds + non-negative offset is nuw), so annotating this in the frontend is not strictly necessary.

**Comment by rjmccall** [comment]
`CGBuilder` is just aiming to provide an `Address` overload of LLVM's `CreateStructGEP`.  It seems wrong for two overloads to have subtly different behavior.

Is there a reason why LLVM's `CreateStructGEP` doesn't add `nuw` itself?

**Comment by hazzlim** [comment]
> `CGBuilder` is just aiming to provide an `Address` overload of LLVM's `CreateStructGEP`. It seems wrong for two overloads to have subtly different behavior.
> 
> Is there a reason why LLVM's `CreateStructGEP` doesn't add `nuw` itself?

Good point - I can see that it makes more sense to add the flag there.

I can rework this patch to do this. Before doing so - do people think it still makes sense to do add the flag in ‘llvm::CreateStructGEP’, given that as @nikic points out we can infer the flag at the IR level?

**Comment by hazzlim** [comment]
> > `CGBuilder` is just aiming to provide an `Address` overload of LLVM's `CreateStructGEP`. It seems wrong for two overloads to have subtly different behavior.
> > Is there a reason why LLVM's `CreateStructGEP` doesn't add `nuw` itself?
> 
> Good point - I can see that it makes more sense to add the flag there.
> 
> I can rework this patch to do this. Before doing so - do people think it still makes sense to do add the flag in ‘llvm::CreateStructGEP’, given that as @nikic points out we can infer the flag at the IR level?

@rjmccall I've reworked this patch so that the nuw flag is added in LLVM's `CreateStructGEP` directly.

**Comment by rjmccall** [comment]
Well, that's a heroic amount of test updates.

I have no problem with the clang changes here.  I'm happy to leave it up to middle-end folks whether adding this attribute is worthwhile given that it can be trivially derived from `inbounds`.

File: llvm/include/llvm/IR/IRBuilder.h
Comment (nikic) [inline llvm/include/llvm/IR/IRBuilder.h:1913]: Shouldn't this be?
```suggestion
            Folder.FoldGEP(Ty, Ptr, Idxs, NWFlags))
```

**Review by nikic — COMMENTED** [review]
As you already put in the test update work, I'm happy to take it :)

File: llvm/include/llvm/IR/IRBuilder.h
Comment (hazzlim) [reply llvm/include/llvm/IR/IRBuilder.h]: Ah yes good point - thanks for spotting!

I've updated with this change, as well as a couple of arch specific test updates that were missed before due to my build config.

Comment (nikic) [reply llvm/include/llvm/IR/IRBuilder.h]: It looks like this one still passes `NWFlags.isInBounds()` instead of the full `NWFlags`?

Comment (hazzlim) [reply llvm/include/llvm/IR/IRBuilder.h]: Yes, sorry about that - updated to pass the full `NWFlags` correctly (+ associated test updates).

**Review by nikic — APPROVED** [review]
LGTM once CI is happy. It looks like there are some mlir test failures.

**Pull request merged**
